src/journal/compress.c

```diff
@@ -667,13 +677,237 @@ int decompress_stream_lz4(int in, int out, uint64_t max_bytes) {
         return -EPROTONOSUPPORT;
 #endif
 }
+#if HAVE_ZSTD
+static int zstd_ret_to_errno(size_t ret) {
+        static_assert(ZSTD_VERSION_NUMBER >= 10301,
+                      "<zstd_errors.h> is stable since zstd 1.3.1");
+
+        ZSTD_ErrorCode error = ZSTD_getErrorCode(ret);
+
+        switch (error) {
+        case ZSTD_error_dstSize_tooSmall:
+                return -ENOBUFS;
+        case ZSTD_error_memory_allocation:
+                return -ENOMEM;
+        default:
+                return -EBADMSG;
+        }
+}
+#endif
+
+int compress_stream_zstd(int fdf, int fdt, uint64_t max_bytes) {
+#if HAVE_ZSTD
+        _cleanup_(ZSTD_freeCCtxp) ZSTD_CCtx *cctx = NULL;
+        _cleanup_free_ void *buffIn = NULL;
+        _cleanup_free_ void *buffOut = NULL;
+        size_t buffInSize, buffOutSize;
+        size_t z;
+        uint64_t left = max_bytes, in_bytes = 0;
+        uint64_t inputSize = 0;
+
+        assert(fdf >= 0);
+        assert(fdt >= 0);
+
+        /* Create the context and buffers */
+        buffInSize = ZSTD_CStreamInSize();
+        buffOutSize = ZSTD_CStreamOutSize();
+        buffIn = malloc(buffInSize);
+        buffOut = malloc(buffOutSize);
+        cctx = ZSTD_createCCtx();
+        if (!cctx || !buffOut || !buffIn)
+                return -ENOMEM;
+
+        if (inputSize) {
+                z = ZSTD_CCtx_setPledgedSrcSize(cctx, inputSize);
+                if (z)
+                        log_debug("Failed to enable Zstd Inputsize: %s",
+                                ZSTD_getErrorName(z));
+        }
+
+        z = ZSTD_CCtx_setParameter(cctx, ZSTD_c_checksumFlag, 1);
+        if (z)
+                log_debug("Failed to enable Zstd Checksum: %s",
+                          ZSTD_getErrorName(z));
+
+        /* This loop read from the input file, compresses that entire chunk,
+         * and writes all output produced to the output file.
+         */
+        for (bool lastChunk = false; !lastChunk;) {
+                ZSTD_inBuffer input = {.src = buffIn, .size = 0, .pos = 0};
+                ssize_t red;
+
+                red = loop_read(fdf, buffIn, buffInSize, 1);
+                if (_unlikely_(red <= 0)) {
+                        if (red < 0)
+                                return -errno;
+                        lastChunk = true;
+                }
+
+                in_bytes += (size_t)red;
+                input.size = (size_t)red;
+
+                for (bool finished = false; !finished;) {
+                        ZSTD_outBuffer output = {
+                            .dst = buffOut, .size = buffOutSize, .pos = 0};
```

Comment:
strange indenting. please format as: 

 ```c
ZSTD_outBuffer output = {
        .dst = buffOut,
        .size = buffOutSize,
        .pos = 0
};
```

(here and elswhere)

Reply:
I did a reformat with 100 chars per line (finally a clang-format in upstream), means those now look like this:
```
ZSTD_outBuffer output = { .dst = out_buff, .size = out_allocsize, .pos = 0 };
```

That's ok too?
This did fix a few more style inconsistencies like adding a space after a cast.

Reply:
i much prefer the multi-line version. it's so much more readable than cramping everyting in one line...

(but this doesn't really matter much)